engine.go

```diff
@@ -1,222 +1,294 @@
 package fished
 
 import (
+	"runtime"
 	"sync"
+	"time"
 
 	"github.com/knetic/govaluate"
+	"github.com/patrickmn/go-cache"
 )
 
-// Config ...
-type Config struct {
-	Worker         int
-	WorkerPoolSize int
-	Cache          bool
-}
+var (
+	//DefaultTarget is the default target facts
+	DefaultTarget = "result_end"
 
-// Engine ...
-type Engine struct {
-	Rules        []Rule
-	rf           map[string]govaluate.ExpressionFunction
-	Jobs         chan int
-	Config       *Config
-	work         sync.WaitGroup
-	wmMutex      sync.RWMutex
-	runMutex     sync.Mutex
-	factsMutex   sync.RWMutex
-	err          []error
-	wm           map[string]interface{}
-	workingRules map[string][]int
-}
+	// DefaultWorker is the default worker for Engine
+	DefaultWorker = 0
+)
 
-// Rule ...
-type Rule struct {
-	Output      string   `json:"output"`
-	Input       []string `json:"input"`
-	Expression  string   `json:"expression"`
-	ee          *govaluate.EvaluableExpression
-	facts       map[string]interface{}
-	result      interface{}
-	hasExecuted bool
-}
+type (
+	// Engine core of the machine
+	Engine struct {
+		InitialFacts  map[string]interface{}
+		Rules         []Rule
+		RuleFunctions map[string]govaluate.ExpressionFunction
+		RuleCache     *cache.Cache
+		RunLock       sync.RWMutex
+	}
 
-// RuleRaw ...
-type RuleRaw struct {
-	Data []Rule `json:"data"`
-}
+	// Rule is struct for rule in fished
+	Rule struct {
+		Input      []string `json:"input"`
+		Output     string   `json:"output"`
+		Expression string   `json:"expression"`
+	}
 
-// RuleFunction ...
-type RuleFunction func(arguments ...interface{}) (interface{}, error)
+	// RuleFunction if type defined for rule function
+	RuleFunction func(...interface{}) (interface{}, error)
 
-var workerPoolSize = 10
+	// Runtime is an struct for each time Engine.Run() is called
+	Runtime struct {
+		Facts      map[string]interface{}
+		JobCh      chan *Job
```

Comment:
I'm guessing variable name suffixed with `Ch` would stands for channel?

Reply:
I guess it's common pattern I will use in the future to name suffix the vars with `Ch`

Reply:
Yes correct